# Examen Final

- Prender cámaras.
- Puede utilizar sus apuntes.
- Puede utilizar los notebooks del curso.
- La alumna o alumno que sea sorprendido solicitando ayuda a cualquier persona, sea ella compañera o no, será calificada con un 2.
- Cada pregunta vale un punto.
- La nota mínima es un 2.
- La nota se determinará con la siguiente fórmula $nota = \min\left(puntos + 2, 7\right)$
- Todas las funciones necesarias para resolver los ejercicios están ya importadas al notebook.

## Librerías

In [5]:
from finrisk import examen_final as ef
from scipy.interpolate import interp1d
import modules.hull_white as hw
import pandas as pd
import numpy as np
import textwrap
import random
import math

## Curva Cero Cupón

Los plazos están en días y las tasas en convención exp act/365.

In [6]:
df_curva = pd.read_excel('data/20201012_built_sofr_zero.xlsx')

In [7]:
df_curva.head()

plazo      tasa        df
0      1  0.000811  0.999998
1      7  0.000841  0.999984
2     14  0.000780  0.999970
3     21  0.000774  0.999955
4     33  0.000781  0.999929

In [8]:
curva = interp1d(df_curva['plazo'], df_curva['tasa'], 'linear', fill_value='extrapolate')

## Cálculo Opción

Para identificar si es una Call o una Put se usa un `enum`.

In [9]:
c_p = hw.CallPut.CALL
c_p = hw.CallPut.PUT

Ver toda la documentación:

In [10]:
print(hw.zcb_call_put.__doc__)


    Calcula el valor de una call o una put sobre un bono cero cupón en el modelo de HW.
    
    params:
    
    - c_p: indica si es la opción es Call o Put. Es un `enum` de tipo CallPut. Ejemplo, c_p = CallPut.CALL.
    - strike: es el strike de la opción. Se ingresa como número. Un strike del 90% se ingresa como .9.
    - r0: es la tasa corta al momento de valorizar la opción (t = 0).
    - to: instante de tiempo en que vence la opción, expresado en años.
    - tb: instante de tiempo en que vence el bono subyacente, expresado en años. Debe ser tb > to.
    - zo: factor de descuento de mercado a tiempo t = 0 hasta to.
    - zb: factor de descuento de mercado a tiempo t = 0 hasta tb.
    - gamma: parámetro gamma del modelo HW.
    - sigma: parámetro sigma del modelo HW.
    
    return:
    
    -  el valor de la opción.
    


## Valores `gamma` y `sigma`

In [11]:
gamma = 1
sigma = .005

In [12]:
strikes = [.98, .981, .982, .983, .984, .985, .99, .991, .992, .993]
for k in strikes:
    print(k, hw.zcb_call_put(
        hw.CallPut.CALL,
        k,
        curva(.0001),
        1,
        2,
        math.exp(-curva(365)),
        math.exp(-curva(730) * 730 / 365.0),
        gamma,
        sigma
    ))

0.98 0.019553575799963863
0.981 0.0185542779311344
0.982 0.017554980062305048
0.983 0.016555682193475696
0.984 0.015556384324646233
0.985 0.014557086455816881
0.99 0.0095605971116699
0.991 0.008561299242840548
0.992 0.007562001374011196
0.993 0.006562703505181733


## Preguntas

In [13]:
try:
    q = ef.get_questions()
except Exception as e:
    print(str(e))

In [14]:
for qq in q:
    print(f'{qq[0]}\n')
    print(textwrap.fill(f'{qq[1]}', 80))
    print('\n')

Pregunta 1:

Calcule la TNA con ICP0 = 10,000.00 e ICP365 = 10,033.00 (365 días después).


Pregunta 2:

Explique la diferencia entre valorizar un payoff g(r(T), T) por simulación de
Montecarlo utilizando el modelo de Hull-White en la medida libre de riesgo y el
modelo de Hull-White en la medida T-forward.


Pregunta 3:

Calcule un paso de simulación de Montecarlo para el modelo de Vasicek usando
gamma = 1.0, sigma = 0.5%, dt = 1/264, r0 obtenido de la curva cupón cero
entregada y número aleatorio N(0, 1) igual a .1.4756 .


Pregunta 4:

Usando la curva cero cupón entregada y parámetros gamma = 1.0 y sigma = 0.5%
valorice una call a 1Y sobre un bono cero cupón a 2Y con un strike de 98.5%.
Entregue el resultado 8 decimales y suponiendo un nocional de 1.


Pregunta 5:

Ceteris paribus una opción sobre una activo con reversión a la media vale menos
que una opción sobre un activo sin reversión a la media. Comente.


Pregunta 6:

Considere un swap a 1Y con cupones semestrales de 1.4% (linea

## Respuestas

Pregunta 1:

Calcule la TNA con ICP0 = 10,000.00 e ICP365 = 10,033.00 (365 días después).


Pregunta 2:

Explique la diferencia entre valorizar un payoff g(r(T), T) por simulación de
Montecarlo utilizando el modelo de Hull-White en la medida libre de riesgo y el
modelo de Hull-White en la medida T-forward.


Pregunta 3:

Calcule un paso de simulación de Montecarlo para el modelo de Vasicek usando
gamma = 1.0, sigma = 0.5%, dt = 1/264, r0 obtenido de la curva cupón cero
entregada y número aleatorio N(0, 1) igual a .1.4756 .


Pregunta 4:

Usando la curva cero cupón entregada y parámetros gamma = 1.0 y sigma = 0.5%
valorice una call a 1Y sobre un bono cero cupón a 2Y con un strike de 98.5%.
Entregue el resultado 8 decimales y suponiendo un nocional de 1.


Pregunta 5:

Ceteris paribus una opción sobre una activo con reversión a la media vale menos
que una opción sobre un activo sin reversión a la media. Comente.


Pregunta 6:

Considere un swap a 1Y con cupones semestrales de 1.4% (lineal). Si la tasa
cupón cero a 0.5Y es 1.3% (exp), calcule la tasa cero a 1Y.


Pregunta 7:

Con la curva cero cupón, interpolando linealmente en tasa, calcule la tasa
forward entre 1282 y 2454 días.


### Pregunta 1

In [8]:
ICP0 = 10000.00
ICP365 = 10033.00
d = 365
TNA = round((ICP365 / ICP0 -1)* 360 / d,4 )
print(f'{TNA:.8%}')

0.33000000%


### Pregunta 2

La diferencia principal radica en que la medida forward se utiliza cuando debemos considerar el plazo de un instrumento a valorizar, en este caso, la medida forward por otra parte modifica el factor de descuento llevandolo fuera de la ecuación, asi como, a diferencia del modelo con la medida riesgo neutral, cambia el theta.

### Pregunta 3

In [37]:
r_ = 0.005
gamma = 1.0
sigma = 0.005
dt = 1/264
randcero = .14756
r0 = df_curva.iloc[0]['tasa']
dt_gamma_r_ = dt * gamma * r_
sigma_sqdt = sigma * math.sqrt(dt)
result = [(0, r0),]
gamma_dt = gamma * dt
r=r0
print(f'{r:.8%}')
r = dt_gamma_r_ + (1 - gamma_dt) * r + sigma_sqdt * randcero
print(f'{r:.8%}')

0.08111102%
0.08723857%


### Pregunta 4

In [22]:
gamma = 1
sigma = .005
strikes = [ .985]
for k in strikes:
        Call=hw.zcb_call_put(
        hw.CallPut.CALL,
        k,
        curva(.0001),
        1,
        2,
        math.exp(-curva(365)),
        math.exp(-curva(730) * 730 / 365.0),
        gamma,
        sigma
    )
print(k,f'{Call:.8%}')

0.985 1.45570865%


### Pregunta 5

Verdadero, incluso bajo la misma volatilidad, la opcion con reversión a la media vale menos, porque la reversion a la media disminuye la volatilidad realizada, aun bajo un mismo sigma.

### Pregunta 6

In [7]:

r=0.014
cz06=0.013
T1=180
T2=360
df2 = math.exp(-cz06*T1 /365)
cz12=(r * (T2 - T1) /360 + 1) / df2
cz12 = 1 / cz12
cz12= math.log(cz12)*(-1)*365 / T2
print(f'{cz12:.8%}')

1.35724973%


### Pregunta 7

In [15]:
t1 = 1282
t2 = 2454
r1 = curva(t1)
r2 = curva(t2)
df1 = math.exp(-r1*t1 /365)
df2 = math.exp(-r2*t2 /365)
fwd12 = (df1 / df2 - 1.0) * 360.0 / (t2 - t1)
print(f'{fwd12:.8%}')

0.59331987%
